### PR TITLE
fix(sdk): make workflow validation compatible with argo CLI v2.12+. Fixes #4775, Fixes #5296

### LINKED ISSUE #4775: Validating a compiled Kubeflow Pipeline with Argo v2.7+ fails
### What steps did you take:

Since KFP [upgraded to Argo==2.7](https://github.com/kubeflow/pipelines/pull/3537), looks like the validation of compiled pipelines no longer works, see [this line](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/compiler/compiler.py#L1000)

Running `argo lint /dev/stdin ''` will always fail now whereas looks like that was used as a check in `argo ~2.4`.

### What happened:

It errors when trying to validate a workflow.

### What did you expect to happen:

I expected it to validate the workflow.

### Environment:

KFP SDK version: v1.0.4 (actually a version of `master` from a few weeks back).

### Anything else you would like to add:

Ideally it'd be great to be able to disable validation via the `compile` API/CLI. For now I've just removed `argo` from the `PATH` but obviously not ideal.

/kind bug
/area sdk

### LINKED ISSUE #5296: [sdk] KFP SDK fail to use argo CLI >=2.12 to lint workflows
### What happened:
After https://github.com/kubeflow/pipelines/pull/5266, I noticed in tests, the warning

> Cannot validate the compiled workflow. Found the argo program in PATH, but it's not usable. argo v2.4.3 should work.

### What did you expect to happen:

KFP SDK can use latest argo CLI to validate workflows locally.

### Environment:

KFP SDK version: 1.4.0

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

/area sdk

The root cause seems to be that, KFP SDK tries to lint empty input using argo CLI to verify if argo CLI works.

However, the default behavior of Argo CLI changed, it gives an error when validating empty string.

We should update to validate an empty workflow instead.

See https://github.com/kubeflow/pipelines/blob/21e3ef8311308c748c0b1656a5f8a06df0ea3045/sdk/python/kfp/compiler/compiler.py#L1094

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**
Fixes #5296

Update the validation of argo cli to be a small but a correct workflow. Open for discussion if we really like to validate with a workflow, but an empty string is not supported by argo lint any longer. 

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Ark-kun** [comment]
Thank you for the fix.
/lgtm
/approve

**Comment by Ark-kun** [comment]
@NikeNano Does Argo 2.7.5 CLI really work well? AFAIK, it lacks "offline" mode and cannot run without a cluster.

**Comment by Bobgy** [comment]
and we are still using 2.4.3 cli: https://github.com/kubeflow/pipelines/blob/9597c8b4f49678163c263601ad73c8d48d779b70/backend/Dockerfile#L19

**Comment by NikeNano** [comment]
@Bobgy , @Ark-kun  how do we like to do about the future of the local validation? Should we check if a cluster is available and then do it or remove it?

**Comment by Bobgy** [comment]
I don't fully follow, the issue https://github.com/argoproj/argo/issues/2793 seems fixed and closed. Is argo client still not supporting local linting?

**Comment by NikeNano** [comment]
I should have linked to these as well: 
- https://github.com/argoproj/argo/issues/1662
- https://github.com/argoproj/argo/pull/2848
 instead to make it more clear @Bobgy. 

The [documentation](https://argoproj.github.io/argo/cli/argo_lint/) also mentions:
```
--kubeconfig string              Path to a kube config. Only required if out-of-cluster
```

The issue by @Ark-kun that I linked to takes a wider scope, and the cli should now work with out a kubeconfig, but the linting I believe will not work.

**Comment by NikeNano** [comment]
@Bobgy any thought on this, otherwise I will close this PR until it is updated upstreams. Unclear to me if it is better to stick to 2.4.3 locally or nothing at all.

**Comment by Bobgy** [comment]
@NikeNano I think we can stick to 2.4.3 until we find a real problem with it.

**Comment by NikeNano** [comment]
> @NikeNano I think we can stick to 2.4.3 until we find a real problem with it (while waiting for upstream updates).

Ok, then I close this.

**Comment by Bobgy** [comment]
Ohh, I see. Thanks! the fix is spot on!

Would you mind fixing the merge conflict?

**Comment by NikeNano** [comment]
This still seems to be an issue for argo 2.12.10
``` bash
docker run -it docker/whalesay:latest sh
apt-get update
apt-get install -y curl

curl -sLO https://github.com/argoproj/argo/releases/download/v2.12.10/argo-linux-amd64.gz
gunzip argo-linux-amd64.gz
chmod +x argo-linux-amd64
mv ./argo-linux-amd64 /usr/local/bin/argo
argo version
argo --loglevel debug lint
```

**Comment by Bobgy** [comment]
https://github.com/argoproj/argo-workflows/issues/2793#issuecomment-799725886

Looks like we can only revert back to 2.4 for now

**Comment by Ark-kun** [comment]
/lgtm
/hold
The tests seem to work. We can probably merge this compiler validation improvement. Remove hold as needed.

**Comment by Bobgy** [comment]
I was checking this as well, took a look at logs and they seem good to me.

**Comment by Bobgy** [comment]
Hmm, actually when I disable my laptop's network. I can still use argo lint to verify a local workflow. Therefore, I guess the only problem was when the workflow refers to other argo Kubernetes resources like ClusterWorkflowTemplate, WorkflowTemplate, which KFP does not use at all.

> argo --loglevel debug lint workflow.yaml
DEBU[2021-03-17T17:11:44.704Z] CLI version                                   version="{v2.12.9 2021-02-16T22:46:40Z 737905345d70ba1ebd566ce1230e4f971993dfd0 v2.12.9 clean go1.13 gc darwin/amd64}"
DEBU[2021-03-17T17:11:44.704Z] Client options                                opts="(argoServerOpts=(url=,path=,secure=false,insecureSkipVerify=false,http=false),instanceID=)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template by name
DEBU[2021-03-17T17:11:44.714Z] Resolving the template                        base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template                          base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template by name                  base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template by name                  base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Resolving the template                        base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template                          base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
DEBU[2021-03-17T17:11:44.714Z] Getting the template by name                  base="*v1alpha1.Workflow (namespace=kubeflow,name=)" depth=0 tmpl="*v1alpha1.WorkflowStep (whalesay)"
workflow.yaml is valid
Workflow manifests validated

**Comment by Bobgy** [comment]
/hold
I tried again in docker, it actually fails.

OK, I seem to have a better understanding. Argo CLI tries to authenticate using local kubconfig, it fails if it couldn't do so.
It will never try to connect to kubernetes api if not necessary.

**Comment by Bobgy** [comment]
@NikeNano because of challenges on the argo side, I was just thinking, it might be easier to run `argo lint` on KFP api server, because there's always kubeconfig there.

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -1091,10 +1091,20 @@ def _validate_workflow(workflow: dict):
...
+      spec: 
+        entrypoint: whalesay 
+        templates: 
+        - name: whalesay 
+          container: 
+            image: docker/whalesay:latest""")
     except:
-      warnings.warn("Cannot validate the compiled workflow. Found the argo program in PATH, but it's not usable. argo v2.4.3 should work.")
-
+      warnings.warn("Cannot validate the compiled workflow. Found the argo program in PATH, but it's not usable. argo v2.12.10 should work.")
```
Comment (Bobgy) [inline sdk/python/kfp/compiler/compiler.py]: For this PR, when people see this error message, I think we should let me keep using v2.4.3.

However, if they do have a kubeconfig, they can simply use later versions of argo CLI.

**Comment by Bobgy** [comment]
@Ark-kun @NikeNano what do you think about the proposed solution?

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (NikeNano) [reply sdk/python/kfp/compiler/compiler.py]: Sounds good to me, I will check the work required in argo to make it work locally without a kubeconfig as well.

**Comment by NikeNano** [comment]
> @Ark-kun @NikeNano what do you think about the proposed solution?

LGTM

**Comment by NikeNano** [comment]
> @NikeNano because of challenges on the argo side, I was just thinking, it might be easier to run `argo lint` on KFP api server, because there's always kubeconfig there. It can also return useful information to users.
> 
> EDIT: this can be a future todo.

Yeah, make sense to me.

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -1091,10 +1091,20 @@ def _validate_workflow(workflow: dict):
...
+      spec: 
+        entrypoint: whalesay 
+        templates: 
+        - name: whalesay 
+          container: 
+            image: docker/whalesay:latest""")
     except:
       warnings.warn("Cannot validate the compiled workflow. Found the argo program in PATH, but it's not usable. argo v2.4.3 should work.")
-
+
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py:1107]: Not: Whitespace...

File: sdk/python/kfp/compiler/compiler.py [hunk 3]
```diff
@@ -1091,10 +1091,20 @@ def _validate_workflow(workflow: dict):
   if argo_path:
     has_working_argo_lint = False
     try:
-      has_working_argo_lint = _run_argo_lint('')
+      has_working_argo_lint = _run_argo_lint("""
+      apiVersion: argoproj.io/v1alpha1
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py:1095]: Can we indent 1 more level?

**Comment by Ark-kun** [comment]
> @NikeNano because of challenges on the argo side, I was just thinking, it might be easier to run `argo lint` on KFP api server, because there's always kubeconfig there. It can also return useful information to users.

I think we have an Issue to enable submission-time pipeline validation.
But compile-time validation is useful regardless of the backend validation.

Is there any problem if we merge this PR? It seems like a strict improvement regardless of argo version.

**Comment by Bobgy** [comment]
Yes, I think so. It is a strict improvement in all cases.

**Comment by Bobgy** [comment]
@Ark-kun thank you for the comments, I fixed them. I assume this is good to go.
/lgtm
/approve

**Pull request merged**